spring-cloud-function-adapters/spring-cloud-function-adapter-gcp/src/test/java/org/springframework/cloud/function/adapter/gcloud/integration/LocalServerTestSupport.java

```diff
@@ -0,0 +1,190 @@
+
+			HttpHeaders headers = new HttpHeaders();
+
+			ResponseEntity<String> response = testRestTemplate.postForEntity(
+					"http://localhost:" + serverProcess.getPort(), new HttpEntity<>(gson.toJson(input), headers),
+					String.class);
+
+			assertThat(response.getBody()).isEqualTo(gson.toJson(expectedOutput));
+		}
+		catch (Exception e) {
+			e.printStackTrace();
+		}
+	}
+
+	private static ServerProcess startServer(Class<?> springApplicationMainClass, String function)
+			throws InterruptedException, IOException {
+		int port = nextPort;
+		nextPort = port + 1;
+
+		String signatureType = "http";
+		String target = FunctionInvoker.class.getCanonicalName();
+
+		File javaHome = new File(System.getProperty("java.home"));
+		assertThat(javaHome.exists()).isTrue();
+		File javaBin = new File(javaHome, "bin");
+		File javaCommand = new File(javaBin, "java");
+		assertThat(javaCommand.exists()).isTrue();
+		String myClassPath = System.getProperty("java.class.path");
+		assertThat(myClassPath).isNotNull();
+
+		List<String> command = new ArrayList<>();
+		command.addAll(Arrays.asList(javaCommand.toString(), "-classpath", myClassPath, Invoker.class.getName()));
+
+		ProcessBuilder processBuilder = new ProcessBuilder().command(command).redirectErrorStream(true);
+		Map<String, String> environment = new HashMap<>();
+		environment.put("PORT", String.valueOf(port));
+		environment.put("K_SERVICE", "test-function");
+		environment.put("FUNCTION_SIGNATURE_TYPE", signatureType);
+		environment.put("FUNCTION_TARGET", target);
+		environment.put("MAIN_CLASS", springApplicationMainClass.getCanonicalName());
+		if (function != null) {
+			environment.put("spring.cloud.function.definition", function);
+		}
+		processBuilder.environment().putAll(environment);
+		Process serverProcess = processBuilder.start();
+		CountDownLatch ready = new CountDownLatch(1);
+		StringBuilder output = new StringBuilder();
+		Future<?> outputMonitorResult = EXECUTOR
+				.submit(() -> monitorOutput(serverProcess.getInputStream(), ready, output));
+		boolean serverReady = ready.await(5, TimeUnit.SECONDS);
+		if (!serverReady) {
+			serverProcess.destroy();
+			throw new AssertionError("Server never became ready");
+		}
+		return new ServerProcess(serverProcess, outputMonitorResult, output, port);
+	}
+
+	private static void monitorOutput(InputStream processOutput, CountDownLatch ready, StringBuilder output) {
+		try (BufferedReader reader = new BufferedReader(new InputStreamReader(processOutput))) {
+			String line;
+			while ((line = reader.readLine()) != null) {
+				if (line.contains(SERVER_READY_STRING)) {
+					ready.countDown();
+				}
+				System.out.println(line);
+				synchronized (output) {
+					output.append(line).append('\n');
+				}
+				if (line.contains("WARNING")) {
+					throw new AssertionError("Found warning in server output:\n" + line);
+				}
+			}
+		}
+		catch (IOException e) {
+			e.printStackTrace();
+			throw new UncheckedIOException(e);
```

Comment:
If an exception is thrown here, does it later mean that `serverReady == false` or does it throw an exception which kills the test? If it kills the test then the server process created earlier might not be destroyed.

Reply:
I think it will just become `serverReady == false`.